lib/pyfrc/physics/tankmodel.py

```diff
@@ -0,0 +1,368 @@
+
+            class PhysicsEngine:
+                
+                def __init__(self, physics_controller):
+                    self.physics_controller = physics_controller
+                    
+                    self.drivetrain = tankmodel.TankModel.theory(motors.MOTOR_CFG_CIM_IMP,
+                                                                 robot_mass=90 * units.lbs,
+                                                                 gearing=10.71, nmotors=2,
+                                                                 x_wheelbase=2.0*feet,
+                                                                 wheel_diameter=6*units.inch)
+                    
+                def update_sim(self, hal_data, now, tm_diff):
+                    # TODO: get motor values from hal_data
+                    velocity, rotation = self.drivetrain.get_vector(l_motor, r_motor)
+                    self.physics_controller.drive(velocity, rotation, tm_diff)
+                    
+                    # optional: compute encoder
+                    # l_encoder = self.drivetrain.l_position * ENCODER_TICKS_PER_FT
+                    # r_encoder = self.drivetrain.r_position * ENCODER_TICKS_PER_FT
+    '''
+    
+    
+    @classmethod
+    def theory(cls,
+               motor_config: MotorModelConfig,
+               robot_mass: units.Quantity,
+               gearing: float,
+               nmotors: int = 1,
+               x_wheelbase: units.Quantity = _kitbot_wheelbase,
+               robot_width: units.Quantity = _kitbot_width,
+               robot_length: units.Quantity = _kitbot_length,
+               wheel_diameter: units.Quantity = 6 * units.inch,
+               vintercept: units.volts = 1.3 * units.volts,
+               timestep: int = 1 * units.ms):
+        r'''
+            Use this to create the drivetrain model when you haven't measured
+            ``kv`` and ``ka`` for your robot.
+            
+            :param motor_config:    Specifications for your motor
+            :param robot_mass:      Mass of the robot
+            :param gearing:         Gear ratio .. so for a 10.74:1 ratio, you would pass 10.74
+            :param nmotors:         Number of motors per side
+            :param x_wheelbase:     Wheelbase of the robot
+            :param robot_width:     Width of the robot
+            :param robot_length:    Length of the robot
+            :param wheel_diameter:  Diameter of the wheel
+            :param vintercept:      The minimum voltage required to generate enough
+                                    torque to overcome steady-state friction (see the
+                                    paper for more details)
+            :param timestep_ms:     Model computation timestep
+            
+            Computation of ``kv`` and ``ka`` are done as follows:
+            
+            .. math::
+            
+                velocity_{max} = \frac{\omega_{free} \cdot \pi \cdot d_{wheels} }{r_{gearing}}
```

Comment:
When introducing variables for the first time in mathematical expressions, one should define them. For example "$d_{wheels}$ is the diameter of the wheels, $\omega_{free}$ is the angular velocity of the wheels with full voltage applied under no load".

Reply:
Thanks for noticing that. Grabbed the definitions from the paper and put them in.